File: src/games/stendhal/server/core/rp/group/GroupManager.java
/* $Id$ */
/***************************************************************************
 *                   (C) Copyright 2003-2010 - Stendhal                    *
 ***************************************************************************
 ***************************************************************************
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 ***************************************************************************/
package games.stendhal.server.core.rp.group;

import java.util.HashSet;
import java.util.Iterator;
import java.util.Set;

import games.stendhal.server.core.events.LoginListener;
import games.stendhal.server.core.events.LoginNotifier;
import games.stendhal.server.core.events.TurnListener;
import games.stendhal.server.core.events.TurnNotifier;
import games.stendhal.server.entity.player.Player;

/**
 * manages player groups
 *
 * @author hendrik
 */
public class GroupManager implements TurnListener, LoginListener {

	private final Set<Group> groups = new HashSet<Group>();

	/**
	 * creates a new GroupManager.
	 */
	public GroupManager() {
		TurnNotifier.get().notifyInSeconds(60, this);
		LoginNotifier.get().addListener(this);
	}

	/**
	 * returns the group the specified player is a member of.
	 *
	 * @param playerName name of player
	 * @return Group or <code>null</code> if the player is not a member of any group
	 */
	public Group getGroup(String playerName) {
		for (Group group : groups) {
			if (group.hasMember(playerName)) {
				return group;
			}
		}
		return null;
	}


	/**
	 * Creates a new group
	 *
	 * @param playerName name of player
	 * @return the new group or <code>null</code> if the group could not be created
	 */
	public Group createGroup(String playerName) {
		if (getGroup(playerName) != null) {
			return null;
		}

		Group group = new Group();
		group.addMember(playerName);
		groups.add(group);
		return group;
	}


	/**
	 * destroys a group
	 *
	 * @param playerName name of player
	 * @return true, if the group was destroyed; false otherwise
	 */
	public boolean destroyGroup(String playerName) {
		Group group = getGroup(playerName);
		if (group == null) {
			return false;
		}
		group.destory();
		groups.remove(group);
		return true;
	}


	/**
	 * handles timeouts and removes empty groups
	 */
	public void clean() {
		Iterator<Group> itr = groups.iterator();
		while (itr.hasNext()) {
			Group group = itr.next();
			group.clean();
			if (group.isEmpty()) {
				itr.remove();
			}
		}
	}

	/**
	 * cleans up
	 */
	@Override
	public void onTurnReached(int currentTurn) {
		clean();
		TurnNotifier.get().notifyInSeconds(60, this);
	}

	/**
	 * tell the reconnecting client if he is in a group
	 */
	@Override
	public void onLoggedIn(Player player) {
		Group group = getGroup(player.getName());
		if (group != null) {
			group.sendGroupChangeEvent(player);
		}
	}
}

File: srcjs/stendhal/data/GroupManager.ts
/***************************************************************************
 *                (C) Copyright 2017-2023 - Faiumoni e. V.                 *
 ***************************************************************************
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU Affero General Public License as        *
 *   published by the Free Software Foundation; either version 3 of the    *
 *   License, or (at your option) any later version.                       *
 *                                                                         *
 ***************************************************************************/

declare var stendhal: any;


export class GroupManager {

	private members: string[] = [];
	private lootmode = "";
	private leader = "";

	/** Singleton instance. */
	private static instance: GroupManager;


	/**
	 * Retrieves singleton instance.
	 */
	static get(): GroupManager {
		if (!GroupManager.instance) {
			GroupManager.instance = new GroupManager();
		}
		return GroupManager.instance;
	}

	/**
	 * Hidden singleton constructor.
	 */
	private constructor() {
		// do nothing
	}

	updateGroupStatus(members?: string, leader?: string, lootmode?: string) {
		if (members) {
			var memberArray = members.substring(1, members.length - 1).split("\t");
			stendhal.data.group.members = {};
			for (var i = 0; i < memberArray.length; i++) {
				stendhal.data.group.members[memberArray[i]] = true;
			}
			stendhal.data.group.leader = leader;
			stendhal.data.group.lootmode = lootmode;
		} else {
			stendhal.data.group.members = [];
			stendhal.data.group.leader = "";
			stendhal.data.group.lootmode = "";
		}
	}
}
